packages/bpk-component-infinite-scroll/src/withInfiniteScroll.js

```diff
@@ -0,0 +1,168 @@
+      this.onScroll = props.onScroll;
+      this.handleIntersection = this.handleIntersection.bind(this);
+      this.handleSeeMoreClick = this.handleSeeMoreClick.bind(this);
+      this.handleKeyPress = this.handleKeyPress.bind(this);
+      // TODO add polyfill for IE
+      this.observer = new IntersectionObserver(
+        (...args) => this.handleIntersection(...args),
+        { threshold: 1 },
+      );
+    }
+
+    componentDidMount() {
+      if (this.sentinel) {
+        this.observer.observe(this.sentinel);
+      }
+    }
+
+    componentDidUpdate() {
+      if (this.sentinel && this.state.index > 0) {
+        this.observer.observe(this.sentinel);
+      }
+    }
+
+    async loadMore() {
+      const { index, elementsToRender } = this.state;
+      const { elementsPerScroll, seeMoreAfter } = this.props;
+      const nextElements = await this.onItemsFetch(index, elementsPerScroll);
+      if (nextElements && nextElements.length > 0) {
+        const nextIndex = index + elementsPerScroll;
+        return {
+          index: nextIndex,
+          elementsToRender: elementsToRender.concat(nextElements),
+          showSeeMore: seeMoreAfter === index / elementsPerScroll,
+        };
+      }
+      this.props.onScrollFinished({
+        totalNumberElements: elementsToRender.length,
+      });
+      return {
+        isListFinished: true,
+      };
+    }
+
+    async handleIntersection(entries) {
+      const entry = entries[0];
+      if (entry.intersectionRatio === 1) {
+        this.props.onScroll({ currentIndex: this.state.index });
+        this.observer.unobserve(this.sentinel);
+        const newState = await this.loadMore();
+        this.setState(newState);
+      }
+    }
+
+    async handleSeeMoreClick(e) {
+      this.props.onClickSeeMore(e);
+      this.setState(await this.loadMore());
+    }
+
+    handleKeyPress(e) {
+      const { key } = e;
+      if (key === 'Enter' || key === ' ') {
+        this.handleSeeMoreClick(e);
+      }
+    }
+
+    render() {
+      const { elementsToRender, isListFinished, showSeeMore } = this.state;
+      const { seeMoreComponent, loadingComponent } = this.props;
+
+      let spinnerOrButton = null;
+
+      if (!isListFinished) {
+        if (showSeeMore) {
+          spinnerOrButton = (
+            <div
+              role="button"
+              tabIndex={0}
+              onKeyPress={this.handleKeyPress}
+              onClick={this.handleSeeMoreClick}
+            >
+              {seeMoreComponent}
+            </div>
+          );
+        } else {
+          spinnerOrButton = (
+            <div
+              ref={spinner => {
+                this.sentinel = spinner;
+              }}
+              className={loadingComponent ? null : STYLES['bpk-sentinel']}
+            >
+              {loadingComponent}
+            </div>
+          );
+        }
+      }
+
+      return (
+        <div>
+          <ComponentToExtend {...this.props} elements={elementsToRender} />
```

Comment:
Seems like we shouldn't be spreading all props here. The props used in this component shouldn't be forwarded to the component being extended. We can do something like this:

```
const { 
elementsPerScroll,
loadingComponent,
onItemsFetch,
onClickSeeMore,
onScroll,
onScrollFinished,
seeMoreComponent,
seeMoreAfter,
...rest
} = this.props
....
<ComponentToExtend {...rest} elements={elementsToRender} />
```

Or maybe we should just not pass any props in (besides elements), users can pass in all the props they need directly to the component.

Reply:
I think it's better to use `omit` from lodash then.

Reply:
import is as `import omit from 'lodash/omit'`

Reply:
added